frame/pov-limit/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+/// computation heavy.
+pub trait RefTimeWaster {
+	/// Wastes some `ref_time`. Receives a counter as an argument.
+	fn waste_ref_time(counter: u32);
+}
+
+/// Interface for wasting `proof_size`. The `waste_proof_size` should increase
+/// the size of the PoV block.
+pub trait PovWaster {
+	/// Wastes some `proof_size`. Receives a counter as an argument.
+	fn waste_proof_size(counter: u32) -> Weight;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Type that implements the `RefTimeWaster` trait.
+		type RefTimeWaster: RefTimeWaster;
+
+		/// Type that implements the `PovWaster` trait.
+		type PovWaster: PovWaster;
+
+		/// Weight information for this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event {
+		/// The computation limit has been updated by root.
+		ComputationLimitSet { compute: Perbill },
+		/// The storage limit has been updated by root.
+		StorageLimitSet { storage: Perbill },
+	}
+
+	#[pallet::storage]
+	pub(crate) type Compute<T: Config> = StorageValue<_, Perbill, ValueQuery>;
+
+	#[pallet::storage]
+	pub(crate) type Storage<T: Config> = StorageValue<_, Perbill, ValueQuery>;
+
+	#[pallet::storage]
+	pub(super) type TrashData<T: Config> = StorageMap<_, Blake2_128Concat, u32, u32, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {
+		pub compute: Perbill,
+		pub storage: Perbill,
+	}
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self {
+			Self { compute: Default::default(), storage: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {
+			Compute::<T>::set(self.compute);
+			Storage::<T>::set(self.storage);
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {
+			assert!(
+				!T::WeightInfo::waste_ref_time(1).ref_time().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+			// TODO this needs <https://github.com/paritytech/substrate/pull/11637>.
+			assert!(
+				!T::WeightInfo::waste_proof_size_none(1).proof_size().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+		}
+
+		fn on_idle(_: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			let mut meter = WeightMeter::from_limit(remaining_weight);
+			if !meter.check_accrue(T::DbWeight::get().reads(2)) {
+				return Weight::zero() // TODO maybe benchmark
+			}
```

Comment:
we probably need some benchmark for this. "empty_on_idle" or something like this.

Reply:
So like write some benchmark that is giving us some weight, as we here will also need to take POV size into account etc.

Reply:
The benchmark for this will only work once we have https://github.com/paritytech/substrate/pull/11637

Reply:
Yes I know, but when you now write a "proper benchmark", it should later work "automatically" when we have POV tracking included.